File: packages/python/elizaos/bootstrap/actions/mute_room.py
from __future__ import annotations

from dataclasses import dataclass, field
from typing import TYPE_CHECKING

from elizaos.types import Action, ActionExample, ActionResult, Content, MemoryType

if TYPE_CHECKING:
    from elizaos.types import HandlerCallback, HandlerOptions, IAgentRuntime, Memory, State


@dataclass
class MuteRoomAction:
    name: str = "MUTE_ROOM"
    similes: list[str] = field(
        default_factory=lambda: [
            "SILENCE_ROOM",
            "QUIET_ROOM",
            "DISABLE_NOTIFICATIONS",
            "STOP_RESPONDING",
        ]
    )
    description: str = (
        "Mute a room to stop responding and receiving notifications. "
        "Use this when you want to stop interacting with a room temporarily."
    )

    async def validate(
        self, runtime: IAgentRuntime, message: Memory, _state: State | None = None
    ) -> bool:
        room_id = message.room_id
        if not room_id:
            return False

        room = await runtime.get_room(room_id)
        if room is None:
            return False

        world_id = room.world_id
        if world_id:
            world = await runtime.get_world(world_id)
            if world and world.metadata:
                muted_rooms = world.metadata.get("mutedRooms", [])
                if str(room_id) in muted_rooms:
                    return False

        return True

    async def handler(
        self,
        runtime: IAgentRuntime,
        message: Memory,
        state: State | None = None,
        options: HandlerOptions | None = None,
        callback: HandlerCallback | None = None,
        responses: list[Memory] | None = None,
    ) -> ActionResult:
        room_id = message.room_id
        if not room_id:
            return ActionResult(
                text="No room specified to mute",
                values={"success": False, "error": "no_room_id"},
                data={"actionName": "MUTE_ROOM"},
                success=False,
            )

        room = await runtime.get_room(room_id)
        if room is None:
            return ActionResult(
                text="Room not found",
                values={"success": False, "error": "room_not_found"},
                data={"actionName": "MUTE_ROOM"},
                success=False,
            )

        room_name = str(room.name) if room.name else "Unknown Room"

        world_id = room.world_id
        if world_id:
            world = await runtime.get_world(world_id)
            if world and world.metadata:
                muted_rooms = list(world.metadata.get("mutedRooms", []))
                room_id_str = str(room_id)

                if room_id_str not in muted_rooms:
                    muted_rooms.append(room_id_str)
                    world.metadata["mutedRooms"] = muted_rooms
                    await runtime.update_world(world)

        await runtime.create_memory(
            content=Content(
                text=f"Muted room: {room_name}",
                actions=["MUTE_ROOM"],
            ),
            room_id=room_id,
            entity_id=runtime.agent_id,
            memory_type=MemoryType.ACTION,
            metadata={"type": "MUTE_ROOM", "roomName": room_name},
        )

        response_content = Content(
            text=f"I have muted {room_name}. I won't respond to messages there.",
            actions=["MUTE_ROOM"],
        )

        if callback:
            await callback(response_content)

        return ActionResult(
            text=f"Muted room: {room_name}",
            values={
                "success": True,
                "muted": True,
                "roomId": str(room_id),
                "roomName": room_name,
            },
            data={
                "actionName": "MUTE_ROOM",
                "roomId": str(room_id),
                "roomName": room_name,
            },
            success=True,
        )

    @property
    def examples(self) -> list[list[ActionExample]]:
        return [
            [
                ActionExample(
                    name="{{name1}}",
                    content=Content(text="Please stop responding in this channel."),
                ),
                ActionExample(
                    name="{{name2}}",
                    content=Content(
                        text="I'll mute this room and won't respond here anymore.",
                        actions=["MUTE_ROOM"],
                    ),
                ),
            ],
        ]


mute_room_action = Action(
    name=MuteRoomAction.name,
    similes=MuteRoomAction().similes,
    description=MuteRoomAction.description,
    validate=MuteRoomAction().validate,
    handler=MuteRoomAction().handler,
    examples=MuteRoomAction().examples,
)

File: packages/rust/src/bootstrap/actions/mute_room.rs
//! MUTE_ROOM action implementation.

use async_trait::async_trait;
use std::collections::HashMap;
use std::sync::Arc;

use crate::error::{PluginError, PluginResult};
use crate::runtime::IAgentRuntime;
use crate::types::{ActionResult, Content, Memory, MemoryType, State};

use super::Action;

/// Action for muting a room.
pub struct MuteRoomAction;

#[cfg_attr(not(target_arch = "wasm32"), async_trait)]
#[cfg_attr(target_arch = "wasm32", async_trait(?Send))]
impl Action for MuteRoomAction {
    fn name(&self) -> &'static str {
        "MUTE_ROOM"
    }

    fn similes(&self) -> &[&'static str] {
        &["SILENCE_ROOM", "QUIET_ROOM", "DISABLE_NOTIFICATIONS", "STOP_RESPONDING"]
    }

    fn description(&self) -> &'static str {
        "Mute a room to stop responding and receiving notifications. \
         Use this when you want to stop interacting with a room temporarily."
    }

    async fn validate(&self, runtime: &dyn IAgentRuntime, message: &Memory) -> bool {
        let Some(room_id) = message.room_id else {
            return false;
        };

        // Check if room exists
        let Ok(Some(room)) = runtime.get_room(room_id).await else {
            return false;
        };

        // Check if not already muted
        if let Some(world_id) = room.world_id {
            if let Ok(Some(world)) = runtime.get_world(world_id).await {
                if let Some(muted) = world.metadata.get("mutedRooms") {
                    if let Some(arr) = muted.as_array() {
                        let room_str = room_id.to_string();
                        if arr.iter().any(|v| v.as_str() == Some(&room_str)) {
                            return false; // Already muted
                        }
                    }
                }
            }
        }

        true
    }

    async fn handler(
        &self,
        runtime: Arc<dyn IAgentRuntime>,
        message: &Memory,
        _state: Option<&State>,
        _responses: Option<&[Memory]>,
    ) -> PluginResult<ActionResult> {
        let room_id = message.room_id.ok_or_else(|| {
            PluginError::InvalidInput("No room specified to mute".to_string())
        })?;

        let room = runtime
            .get_room(room_id)
            .await?
            .ok_or_else(|| PluginError::NotFound("Room not found".to_string()))?;

        let room_name = room.name.clone().unwrap_or_else(|| "Unknown Room".to_string());

        // Update world's muted rooms
        if let Some(world_id) = room.world_id {
            if let Some(mut world) = runtime.get_world(world_id).await? {
                let mut muted: Vec<String> = world
                    .metadata
                    .get("mutedRooms")
                    .and_then(|v| v.as_array())
                    .map(|arr| {
                        arr.iter()
                            .filter_map(|v| v.as_str().map(String::from))
                            .collect()
                    })
                    .unwrap_or_default();

                let room_str = room_id.to_string();
                if !muted.contains(&room_str) {
                    muted.push(room_str);
                    world.metadata.insert(
                        "mutedRooms".to_string(),
                        serde_json::json!(muted),
                    );
                    runtime.update_world(&world).await?;
                }
            }
        }

        // Create memory of the action
        let mut metadata = HashMap::new();
        metadata.insert("type".to_string(), serde_json::json!("MUTE_ROOM"));
        metadata.insert("roomName".to_string(), serde_json::json!(&room_name));

        runtime
            .create_memory(
                Content {
                    text: format!("Muted room: {}", room_name),
                    actions: vec!["MUTE_ROOM".to_string()],
                    ..Default::default()
                },
                Some(room_id),
                Some(runtime.agent_id()),
                MemoryType::Action,
                metadata,
            )
            .await?;

        Ok(ActionResult::success(format!("Muted room: {}", room_name))
            .with_value("success", true)
            .with_value("muted", true)
            .with_value("roomId", room_id.to_string())
            .with_value("roomName", room_name.clone())
            .with_data("actionName", "MUTE_ROOM")
            .with_data("roomId", room_id.to_string())
            .with_data("roomName", room_name))
    }
}

